tuf/src/database.rs

```diff
@@ -370,20 +342,20 @@ impl<D: DataInterchange> Database<D> {
             //     new timestamp metadata file is older than the trusted timestamp metadata file,
             //     discard it, abort the update cycle, and report the potential rollback attack.
 
-            let trusted_timestamp_version = self.trusted_timestamp_version();
-
-            match new_timestamp.version().cmp(&trusted_timestamp_version) {
-                Ordering::Less => {
-                    return Err(Error::VerificationFailure(format!(
-                        "Attempted to roll back timestamp metadata at version {} to {}.",
-                        trusted_timestamp_version,
-                        new_timestamp.version()
-                    )));
-                }
-                Ordering::Equal => {
-                    return Ok(None);
+            if let Some(trusted_timestamp) = &self.trusted_timestamp {
+                match new_timestamp.version().cmp(&trusted_timestamp.version()) {
+                    Ordering::Less => {
+                        return Err(Error::VerificationFailure(format!(
+                            "Attempted to roll back timestamp metadata at version {} to {}.",
+                            trusted_timestamp.version(),
+                            new_timestamp.version()
+                        )));
+                    }
+                    Ordering::Equal => {
+                        return Ok(None);
```

Comment:
The TUF spec actually forbids metadata having a version 0, so the metadata struct constructors error out if the version is ever 0:

* [root](https://github.com/theupdateframework/rust-tuf/blob/develop/tuf/src/metadata.rs#L775)
* [targets](https://github.com/theupdateframework/rust-tuf/blob/develop/tuf/src/metadata.rs#L1858)
* [snapshot](https://github.com/theupdateframework/rust-tuf/blob/develop/tuf/src/metadata.rs#L1432)
* [timestamp](https://github.com/theupdateframework/rust-tuf/blob/develop/tuf/src/metadata.rs#L1164)
* [metadata description in timestamp/snapshot](https://github.com/theupdateframework/rust-tuf/blob/develop/tuf/src/metadata.rs#L1164)

Previously we were being a bit clever by using an illegal value of `0` here instead of an option to represent a missing value. So in this case, we should only exit early here and in the other places I changed this pattern only if we have a trusted metadata. Otherwise we should always finish the upadte process.

It'd probably make sense to instead use https://doc.rust-lang.org/stable/std/num/struct.NonZeroU32.html to make sure this is semantically correct. We can do that later though.